src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs

```diff
@@ -747,6 +747,20 @@ private string GetHostNameForDns()
             return host;
         }
 
+        private static bool AllPercentEncodingsAreValid(string value)
+        {
+            bool invalidPercentEncoding;
+            var invalidPercentPattern = @"%$|%.$|%[^0-9a-fA-F]|%[0-9a-fA-F][^0-9a-fA-F]";
+            invalidPercentEncoding = Regex.IsMatch(value, invalidPercentPattern);
+
+            if (invalidPercentEncoding)
+            {
+                return false;
+            }
+
+            return true;
```

Comment:
Lines 756-761 seem odd. Why not just do this on line 754:

    return !Regex.IsMatch(value, invalidPercentPattern)

You might even consider renaming the helper method HasInvalidPercentEncodings or eliminating the helper method altogether.

    var invalidPercentPattern = @"%$|%.$|%[^0-9a-fA-F]|%[0-9a-fA-F][^0-9a-fA-F]";
    if (Regex.IsMatch(value, invalidPercentPattern))
    {
        // throw
    }

which is basically what we had originally right?

Reply:
Sorry for the wild goose chase with IsHexEncoding... :)

Reply:
I'm fine with moving the logic back into Parse(). It brings us back to where we were before, but I definitely learned stuff along the way, so no harm done!